core/src/main/scala/org/apache/spark/scheduler/TaskSetManager.scala

```diff
@@ -797,11 +801,12 @@ private[spark] class TaskSetManager(
     maybeFinishTaskSet()
   }
 
-  private[scheduler] def markPartitionCompleted(partitionId: Int, taskInfo: TaskInfo): Unit = {
+  private[scheduler] def markPartitionCompleted(partitionId: Int, taskInfo: Option[TaskInfo])
+  : Unit = {
     partitionToIndex.get(partitionId).foreach { index =>
       if (!successful(index)) {
         if (speculationEnabled && !isZombie) {
-          successfulTaskDurations.insert(taskInfo.duration)
+          taskInfo.foreach { info => successfulTaskDurations.insert(info.duration) }
```

Comment:
this is existing logic, but I have a question here: Do we really need to do it? The task is finished by another TSM, it seems unreasonable to update the statistics for launching speculative tasks in this TSM.

Reply:
I think you're right. What's your opinion ? @squito

Reply:
Yeah there was discussion about this in the past, there are arguments for doing it multiple ways.  This was kind of a compromise with something that avoided a bug and was a reasonable change to put in, see more here: https://github.com/apache/spark/pull/21656

Reply:
I see.